test/sync/src/commonMain/kotlin/io/realm/kotlin/test/mongodb/util/AppServicesClient.kt

```diff
@@ -0,0 +1,623 @@
+        @OptIn(InternalSerializationApi::class)
+        private suspend inline fun <reified T : Any> HttpClient.typedRequest(
+            method: HttpMethod,
+            url: String,
+            crossinline block: HttpRequestBuilder.() -> Unit = {}
+        ): T {
+            return this@typedRequest.request<HttpResponse>(url) {
+                this.method = method
+                this.apply(block)
+            }.readText()
+                .let {
+                    Json { ignoreUnknownKeys = true }.decodeFromString(
+                        T::class.serializer(),
+                        it
+                    )
+                }
+        }
+
+        // Default serializer fails with
+        // InvalidMutabilityException: mutation attempt of frozen kotlin.collections.HashMap
+        // on native. Have tried the various workarounds from
+        // https://github.com/Kotlin/kotlinx.serialization/issues/1450
+        // but only one that works is manual invoking the deserializer
+        @OptIn(InternalSerializationApi::class)
+        private suspend inline fun <reified T : Any> HttpClient.typedListRequest(
+            method: HttpMethod,
+            url: String,
+            crossinline block: HttpRequestBuilder.() -> Unit = {}
+        ): List<T> {
+            return this@typedListRequest.request<HttpResponse>(url) {
+                this.method = method
+                this.apply(block)
+            }.readText()
+                .let {
+                    Json { ignoreUnknownKeys = true }.decodeFromString(
+                        ListSerializer(T::class.serializer()),
+                        it
+                    )
+                }
+        }
+
+        suspend fun build(
+            debug: Boolean,
+            baseUrl: String,
+            dispatcher: CoroutineDispatcher,
+        ): AppServicesClient {
+            val adminUrl = baseUrl + ADMIN_PATH
+            // Work around issues on Native with the Ktor client being created and used
+            // on different threads.
+            // Log in using unauthorized client
+            val unauthorizedClient = defaultClient("realm-baas-unauthorized", debug)
+            val loginResponse = unauthorizedClient.typedRequest<LoginResponse>(
+                HttpMethod.Post,
+                "$adminUrl/auth/providers/local-userpass/login"
+            ) {
+                contentType(ContentType.Application.Json)
+                body = mapOf("username" to "unique_user@domain.com", "password" to "password")
+            }
+
+            // Setup authorized client for the rest of the requests
+            // Client is currently being constructured for each network reques to work around
+            // https://github.com/realm/realm-kotlin/issues/480
+            val accessToken = loginResponse.access_token
+            unauthorizedClient.close()
+
+            val httpClient = defaultClient("realm-baas-authorized", debug) {
```

Comment:
Something is off here. We also pass in a httpClient in the constructor?

Reply:
This is actually a factory method inside the companion object. The client initialization requires of performing some requests that are suspendable. We could add some context switch within the constructor, but having a factory method allows the execution context to be defined by the caller.